// @aws-amplify/datastore/lib/sync/processors/errorMaps.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mutationErrorMap = {
    BadRecord: function (error) { return /^Cannot return \w+ for [\w-_]+ type/.test(error.message); },
    ConfigError: function () { return false; },
    Transient: function () { return false; },
    Unauthorized: function () { return false; },
};
exports.subscriptionErrorMap = {
    BadRecord: function () { return false; },
    ConfigError: function () { return false; },
    Transient: function () { return false; },
    Unauthorized: function (givenError) {
        var _a = givenError.error, _b = __read((_a === void 0 ? {
            errors: [],
        } : _a).errors, 1), _c = _b[0], _d = (_c === void 0 ? {} : _c).message, message = _d === void 0 ? '' : _d;
        var regex = /Connection failed.+Unauthorized/;
        return regex.test(message);
    },
};
exports.syncErrorMap = {
    BadRecord: function (error) { return /^Cannot return \w+ for [\w-_]+ type/.test(error.message); },
    ConfigError: function () { return false; },
    Transient: function () { return false; },
    Unauthorized: function () { return false; },
};
function getMutationErrorType(error) {
    return mapErrorToType(exports.mutationErrorMap, error);
}
exports.getMutationErrorType = getMutationErrorType;
function getSubscriptionErrorType(error) {
    return mapErrorToType(exports.subscriptionErrorMap, error);
}
exports.getSubscriptionErrorType = getSubscriptionErrorType;
function getSyncErrorType(error) {
    return mapErrorToType(exports.syncErrorMap, error);
}
exports.getSyncErrorType = getSyncErrorType;
/**
 * Categorizes an error with a broad error type, intended to make
 * customer error handling code simpler.
 * @param errorMap Error names and a list of patterns that indicate them (each pattern as a regex or function)
 * @param error The underying error to categorize.
 */
function mapErrorToType(errorMap, error) {
    var e_1, _a;
    var errorTypes = __spread(Object.keys(errorMap));
    try {
        for (var errorTypes_1 = __values(errorTypes), errorTypes_1_1 = errorTypes_1.next(); !errorTypes_1_1.done; errorTypes_1_1 = errorTypes_1.next()) {
            var errorType = errorTypes_1_1.value;
            var matcher = errorMap[errorType];
            if (matcher(error)) {
                return errorType;
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (errorTypes_1_1 && !errorTypes_1_1.done && (_a = errorTypes_1.return)) _a.call(errorTypes_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return 'Unknown';
}
exports.mapErrorToType = mapErrorToType;
